docs/api/utility-apis.md

```diff
@@ -55,51 +55,112 @@ from any component inside Backstage, including the ones in `@backstage/core`.
 The only requirement is that they are beneath the `AppProvider` in the react
 tree.
 
-## Registering Utility API Implementations
+## Supplying APIs
 
-The Backstage App is responsible for providing implementations for all Utility
-APIs required by plugins. The example app in this repo registers its APIs inside
-[src/apis.ts](/packages/app/src/apis.ts). Here's an example of how to wire up
-the `ErrorApi` inside an app:
+### API Factories
+
+APIs are registered in the form of `ApiFactories`, which encapsulate the process
+of instantiating an API. It is a collection of three things: the `ApiRef` of the
+API to instantiate, a list of all required dependencies, and a factory function
+that returns a new API instance.
+
+For example, this is the default `ApiFactory` for the `ErrorApi`:
 
 ```ts
-import {
-  ApiRegistry,
-  createApp,
-  alertApiRef,
-  errorApiRef,
-  AlertApiForwarder,
-  ErrorApiForwarder,
-  ErrorAlerter,
-  ConfigApi,
-} from '@backstage/core';
-
-const apis = (config: ConfigApi) => {
-  const builder = ApiRegistry.builder();
-
-  // The alert API is a self-contained implementation that shows alerts to the user.
-  const alertApi = builder.add(alertApiRef, new AlertApiForwarder());
-
-  // The error API uses the alert API to send error notifications to the user.
-  builder.add(errorApiRef, new ErrorAlerter(alertApi, new ErrorApiForwarder()));
-  return builder.build();
-};
+createApiFactory({
+  api: errorApiRef,
+  deps: { alertApi: alertApiRef },
```

Comment:
Makes me think... A bit late now, but could this have made into an array? Deps `[aRef, bRef]`, factory `(a, b) =>`. Should be equally typecheckable but doesn't require repeating the name of the instance. Takes a bit more TS magic though right

Reply:
Definitely possible to change this still.

Afaik we'd need TypeScript 4 to be able to implement this. A straight up array parameter type won't work because it creates a union of all the APIs, i.e. `(A | B)[]`, so you need a tuple, aka readonly array. That somewhat works, as in the types can be forwarded from deps to the factory args, but afaik it's not possible to do the type inference that extracts the implementation type from the `ApiRef`, i.e. the `{ [key in keyof T]: ApiRef<T[key]> }` mapping. That's because readonly arrays don't seem to have an index signature, so the `keyof T` | `T[key]` doesn't work. The new variadic tuple types I assume will help out here, but haven't tried them out yet.

Either way I prefer the slightly more verbose but imo less error-prone explicit naming between `deps` and the `factory` parameter with the objects.